### PR TITLE
Replace munkres with lapjv for track assignment

### LINKED ISSUE #5207: MemoryError during import
Trying to import an album results in a MemoryError, the reason seems to be that one of the candidates it finds on discogs is an album with 6666 tracks. Disabling the discogs plugin so this is not a candidate "fixes" it and lets the import proceed. While albums with that many tracks are not the norm, is there something which can be tweaked so beets can handle that?

### Problem

```sh
$ beet -vv import -a SomeAlbum
...
Success. Distance: 0.06
Sending event: albuminfo_received
Candidate: Lord Poozix - Grim And Frostbitten Penguins (29948101)
Computing track assignment...
Traceback (most recent call last):
  File "/usr/local/bin/beet", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1285, in main
    _raw_main(args)
  File "/usr/local/lib/python3.10/site-packages/beets/ui/__init__.py", line 1272, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/local/lib/python3.10/site-packages/beets/ui/commands.py", line 973, in import_func
    import_files(lib, paths, query)
  File "/usr/local/lib/python3.10/site-packages/beets/ui/commands.py", line 943, in import_files
    session.run()
  File "/usr/local/lib/python3.10/site-packages/beets/importer.py", line 340, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/usr/local/lib/python3.10/site-packages/beets/util/pipeline.py", line 446, in run_parallel
    raise exc_info[1].with_traceback(exc_info[2])
  File "/usr/local/lib/python3.10/site-packages/beets/util/pipeline.py", line 311, in run
    out = self.coro.send(msg)
  File "/usr/local/lib/python3.10/site-packages/beets/util/pipeline.py", line 193, in coro
    func(*(args + (task,)))
  File "/usr/local/lib/python3.10/site-packages/beets/importer.py", line 1376, in lookup_candidates
    task.lookup_candidates()
  File "/usr/local/lib/python3.10/site-packages/beets/importer.py", line 660, in lookup_candidates
    autotag.tag_album(self.items, search_ids=self.search_ids)
  File "/usr/local/lib/python3.10/site-packages/beets/autotag/match.py", line 466, in tag_album
    _add_candidate(items, candidates, matched_candidate)
  File "/usr/local/lib/python3.10/site-packages/beets/autotag/match.py", line 372, in _add_candidate
    mapping, extra_items, extra_tracks = assign_items(items, info.tracks)
  File "/usr/local/lib/python3.10/site-packages/beets/autotag/match.py", line 105, in assign_items
    matching = Munkres().compute(costs)
  File "/usr/local/lib/python3.10/site-packages/munkres.py", line 144, in compute
    self.path = self.__make_matrix(self.n * 2, 0)
  File "/usr/local/lib/python3.10/site-packages/munkres.py", line 181, in __make_matrix
    matrix += [[val for j in range(n)]]
  File "/usr/local/lib/python3.10/site-packages/munkres.py", line 181, in <listcomp>
    matrix += [[val for j in range(n)]]
MemoryError
```

This is the discogs release which triggers the bug:

https://www.discogs.com/release/29948101-Lord-Poozix-Grim-And-Frostbitten-Penguins

### Setup

* OS: OpenBSD 7.4
* Python version: Python 3.10.13
* beets version: 1.60
* Turning off plugins made problem go away (yes/no): yes

My configuration (output of `beet config`) is:

```yaml
$ beet config
lyrics:
    bing_lang_from: []
    auto: yes
    bing_client_secret: REDACTED
    bing_lang_to:
    google_API_key: REDACTED
    google_engine_ID: REDACTED
    genius_api_key: REDACTED
    fallback:
    force: no
    local: no
    sources:
    - google
    - musixmatch
    - genius
    - tekstowo
library: /home/johan/.config/beets/library.db
directory: /music

plugins:
- bucket
- chroma
- discogs
- duplicates
- edit
- embedart
- fetchart
- fromfilename
- info
- inline
- lastgenre
- lyrics
- mbsync
- missing
- permissions
- plexupdate
- replaygain
- rewrite
- web
pluginpath: /home/johan/.config/beets/plugins
ignore: .AppleDouble ._* *~ .DS_Store
ignore_hidden: yes
asciify_paths: yes
art_filename: '%asciify{%lower{$album}}'
original_date: yes
permissions:
    dir: 775
    file: 664

replace:
    -\.\.+: '-'
    \.\.+$: ''
    \.$: ''
    \.\.\.: _
    '[\\/]': ''
    ^\.: ''
    '[\x00-\x1f]': ''
    \._: '-'
    ':': ''
    '[<>"\?\*\|\!]': ''
    \s+$: ''
    ^\s+: ''
    \[\]: ''
    '&': _and_
    '#': _
    \,: ''
    \;: ''
    '''': ''
    \(: ''
    \): ''
    \?: ''
    \!: ''
    "\u2026": ''
    \s: _
    _-_: _
    _\._: _
    _+: _
    "\xBF": ''
    '[\xc2\xbf]': ''

import:
    copy: yes
    duplicate_action: keep
    write: yes
    languages: en sv no dk is
item_fields:
    path_format: "if format == 'FLAC':\n        return 'flac'\nreturn 'mp3'\n"
    path_genre: "for p in ['alternative', 'reggae', 'world music']:\n    if p in genre.lower():\n       return 'pop'\nfor l in ['arabic', 'latin', 'merengue', 'reggaeton', 'salsa', 'tango']:\n    if l in genre.lower():\n        return 'latin'\nfor s in ['musical', 'soundtrack', 'soundtracks']:\n    if s in genre.lower():\n        return 'soundtracks'\nfor r in ['hip-hop', 'hip hop', 'r&b', 'rap']:\n    if r in genre.lower():\n        return 'rnb'\nfor i in ['industrial rock', 'country', 'blues', 'jazz']:\n    if i in genre.lower():\n        return 'rock'\nfor m in ['ambient', 'death', 'doom', 'folk', 'grind', 'core', 'punk', 'industrial']:\n    if m in genre.lower():\n        return 'metal'\nfor g in ['classical', 'comedy', 'humour', 'latin', 'metal', 'pop', 'rock']:\n    if g in genre.lower():\n        return g\nif not genre:\n    return 'fixit'\n"
    path_composer: "if not composer_sort:\n        return composer\nreturn composer_sort\n"
album_fields:
    path_bucket: "for g in ['classical', 'metal', 'rock', 'country', 'blues', 'jazz', 'grindcore', 'funeral doom', 'industrial', 'humour']:\n    if g in genre.lower():\n        return g\n"
    path_year: "if original_year == 0000:\n        return year\nreturn original_year\n"
    path_various: "for v in ['musical', 'soundtrack', 'soundtracks']:\n    if v in genre.lower():\n        return 0\nreturn 1\n"
    path_artist: "if not albumartist_sort:\n        return albumartist\nreturn albumartist_sort\n"
bucket:
    extrapolate: yes
    bucket_alpha:
    - a
    - b
    - c
    - d
    - e
    - f
    - g
    - h
    - i
    - j
    - k
    - l
    - m
    - n
    - o
    - p
    - q
    - r
    - s
    - t
    - u
    - v
    - w
    - x
    - y
    - z
    - 0-9
    bucket_alpha_regex:
        a: "^[\xC6]"
        0-9: ^[0-9]
    bucket_year: []

paths:
    default: $path_format/%lower{$path_genre}/%if{$path_bucket,(%lower{%bucket{$path_artist,alpha}})}/%lower{$path_artist}/${year}-%lower{$album}/${track}-%lower{$title}
    comp: $path_format/%lower{$path_genre}/%if{$path_bucket,v}/%if{$path_various,various_artists}/%lower{$path_artist}/${year}-%lower{$album}/${track}-%lower{$title}
    singleton: $path_format/%lower{$path_genre}/%if{$path_bucket,(%lower{%bucket{$artist_sort,alpha}})}/%lower{$artist_sort}/non-album/${track}-%lower{$title}
    genre:Classical: $path_format/%lower{$path_genre}/%if{$path_bucket,(%lower{%bucket{$path_composer,alpha}})}/%lower{$path_composer}/${year}-%lower{$album}/${track}-%lower{$title}
    series::.: $path_format/%lower{$path_genre}/%if{$path_bucket,v}/various_artists/%lower{$series}/${year}-%lower{$album}/${track}-%lower{$title}
web:
    host: REDACTED
    port: REDACTED
    cors: ''
    cors_supports_credentials: no
    reverse_proxy: no
    include_paths: no
    readonly: yes
chroma:
    auto: yes
acoustid:
    apikey: REDACTED
replaygain:
    auto: no
    albumgain: yes
    backend: ffmpeg
    command: /usr/local/bin/ffmpeg
    overwrite: no
    threads: 12
    parallel_on_import: no
    per_disc: no
    peak: 'true'
    targetlevel: 89
    r128: [Opus]
    r128_targetlevel: 84
metalarchives:
    lyrics: yes
    lyrics_search: yes
edit:
    itemfields:
    - album
    - albumartist
    - artist
    - track
    - title
    - year
    albumfields:
    - albumartist
    - album
    - year
    - albumtype
    - genre
    - original_year
    - original_month
    - original_day
    ignore_fields: id path
duplicates:
    album: no
    checksum: ''
    copy: ''
    count: no
    delete: no
    format: ''
    full: no
    keys: []
    merge: no
    move: ''
    path: no
    tiebreak: {}
    strict: no
    tag: ''
discogs:
    apikey: REDACTED
    apisecret: REDACTED
    tokenfile: discogs_token.json
    source_weight: 0.5
    user_token: REDACTED
    separator: ', '
    index_tracks: no
rewrite: {}
embedart:
    maxwidth: 0
    auto: yes
    compare_threshold: 0
    ifempty: no
    remove_art_file: no
    quality: 0
fetchart:
    auto: yes
    minwidth: 0
    maxwidth: 0
    quality: 0
    max_filesize: 0
    enforce_ratio: no
    cautious: no
    cover_names:
    - cover
    - front
    - art
    - album
    - folder
    sources:
    - filesystem
    - coverart
    - itunes
    - amazon
    - albumart
    google_key: REDACTED
    google_engine: REDACTED
    fanarttv_key: REDACTED
    lastfm_key: REDACTED
    store_source: no
    high_resolution: no
    deinterlace: no
    cover_format:
pathfields: {}
lastgenre:
    whitelist: yes
    min_weight: 10
    count: 1
    fallback:
    canonical: no
    source: album
    force: yes
    auto: yes
    separator: ', '
    prefer_specific: no
    title_case: yes
missing:
    count: no
    total: no
    album: no
```

### PR DESCRIPTION
**Pull request by snejus**
## Description

Fixes #5207.

This PR replaces the `munkres` library with `lap` (Linear Assignment Problem solver) for computing optimal track assignments during the auto-tagging process. The main changes are:

- Remove `munkres` dependency and add `lap` and `numpy` dependencies
- Refactor the track assignment code to use `lap.lapjv()` instead of `Munkres().compute()`
- Simplify cost matrix construction using list comprehension
- Move config value reading outside of `track_distance` function to improve performance

The motivation for this change comes from benchmark comparisons showing that LAPJV (implemented in the `lap` library) significantly outperforms the Munkres/Hungarian algorithm for the linear assignment problem. See detailed benchmarks at: https://github.com/berhane/LAP-solvers

The change should provide better performance for track matching, especially with larger albums, while maintaining the same assignment results.

## Testing Notes
- All existing tests pass without modification
- Track assignments produce identical results 
- No behavioral changes in auto-tagging

### CODE REVIEW

File: beets/autotag/match.py [hunk 1]
```diff
@@ -126,21 +127,15 @@ def assign_items(
...
     """
     # Construct the cost matrix.
-    costs: list[list[Distance]] = []
-    for item in items:
-        row = []
-        for track in tracks:
-            row.append(track_distance(item, track))
-        costs.append(row)
-
+    costs = [[float(track_distance(i, t)) for t in tracks] for i in items]
```
Comment (wisp3rwind) [inline beets/autotag/match.py:130]: I'd be curious how much this quadratic steps contributes to the overall runtime (the lap algorithms seem to scale cubically, but the overhead of the Python code in `track_distance` could be significant).

In any case, even if it turns out to matter, looking at the distance computation should probably be a separate PR.

File: beets/autotag/match.py [hunk 2]
```diff
@@ -154,6 +149,10 @@ def track_index_changed(item: Item, track_info: TrackInfo) -> bool:
     return item.track not in (track_info.medium_index, track_info.index)
 
 
+track_length_grace = config["match"]["track_length_grace"].as_number()
+track_length_max = config["match"]["track_length_max"].as_number()
```
Comment (wisp3rwind) [inline beets/autotag/match.py:153]: I think it's common practice in our codebase to load configuration values at the point of use (unless there's a specific reason to do so earlier, I think there might be some plugins that do so on initialization).

Sometimes our tests depend on this (but apparently not here)?

Is there a specific reason for this change? Otherwise, I'd argue to keep it as is and move it back into `track_distance`.

Actually, this probably was a bottleneck of `track_distance`? I think this would be worth a code comment. Also, regarding the point above, maybe this could be moved to `assign_items` and the values passed to `track_distance`?

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
This seems like the right call — with `munkres` being a pure-Python implementation, it's orders of magnitudes slower than the other libraries. Nice find that somebody already did the hard work of benchmarking all these packages!

The new dependencies look fine at a first glance: `numpy` is common (and I wouldn't be surprised if we depend on it already transitively). `lap` and `munkres` have a similar popularity on Github and have both seen updates about a month ago.

I left a small question about on code changes, otherwise this looks good to me; thanks!

File: beets/autotag/match.py (hunk 1 above)
Comment (snejus) [reply beets/autotag/match.py:130]: While working on this I analysed `track_distance` and `Distance` implementations and found that the only significant overhead was the `config` access.

File: beets/autotag/match.py (hunk 2 above)
Comment (snejus) [reply beets/autotag/match.py:153]: Funnily, tests adjust global config so this may actually still work

> Also, regarding the point above, maybe this could be moved to assign_items and the values passed to track_distance?

But as you suggested I'd rather inject these values into the function so I will make this adjustment now.

> Actually, this probably was a bottleneck of track_distance?

Indeed. There's a fair bit of logic that runs to get these values and convert them to a number, so it becomes significant in larger albums. For example, I tested this trying to match my largest album with 136 to a candidate from Discogs that had 6666 tracks: in this case `track_distance` runs `136 * 6666 = 906,576` times.

File: beets/autotag/match.py (hunk 1 above)
Comment (wisp3rwind) [reply beets/autotag/match.py:130]: Nice, so you did all the work already :fire:

File: beets/autotag/match.py (hunk 2 above)
Comment (snejus) [reply beets/autotag/match.py]: `track_distance` is called in 5 places across this module, so injection would have not been a clean solution. Instead, I added two functions that get these values and cached their calls.

Comment (wisp3rwind) [reply beets/autotag/match.py]: The new solution also isn't as clean as I'd like it to be, but on the other hand, if we don't depend on the config being evaluated at the time of the function call, it's probably good enough for now. We can revisit if that changes. Let's merge this as-is.

**Review by wisp3rwind — APPROVED** [review]
Looks good to me :100: (Assuming CI passes)
